l3kernel/l3candidates.dtx

```diff
@@ -2006,6 +2006,113 @@
+  {
+    \sys_if_engine_luatex:T
+    {
+      \luatex_directlua:D
+      {
+        tex.sprint((status.shell_escape~or~os.execute()) .. " ")
+      }
+    }
+    \sys_if_engine_pdftex:T { \the\pdftex_shellescape:D }
+    \sys_if_engine_ptex:T   { \the\pdftex_shellescape:D }
+    \sys_if_engine_uptex:T  { \the\pdftex_shellescape:D }
+    \sys_if_engine_xetex:T  { \the\shellescape }
+  }
+%  \end{macrocode}
+%
+% \begin{macro}[EXP, pTF]{\sys_if_shell:}
+%   Performs a check for whether shell escape is enabled.  This will
+%   return true if either of restricted or unrestircted shell escape
+%   is enabled.
+%    \begin{macrocode}
+\prg_new_conditional:Nnn \sys_if_shell: { p , T , F , TF }
+  {
+    \if_int_compare:w \__sys_shell_escape: = 0 \exp_stop_f:
+      \prg_return_false:
+    \else:
+      \prg_return_true:
+    \fi:
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}[EXP, pTF]{\sys_if_shell_unrestricted:}
+%   Performs a check for whether \emph{unrestricted} shell escape is
+%   enabled.
+%    \begin{macrocode}
+\prg_new_conditional:Nnn \sys_if_shell_unrestricted: { p , T , F , TF }
+  {
+    \if_int_compare:w \__sys_shell_escape: = 1 \exp_stop_f:
+      \prg_return_true:
+    \else:
+      \prg_return_false:
+    \fi:
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}[EXP, pTF]{\sys_if_shell_unrestricted:}
+%   Performs a check for whether \emph{restricted} shell escape is
+%   enabled.  This will return false if unrestricted shell escape is
+%   enabled.  Unrestricted shell escape is not considered a superset
+%   of restricted shell escape in this case.  To find whether any
+%   shell escape is enabled use \cs{sys_if_shell:}.
+%    \begin{macrocode}
+\prg_new_conditional:Nnn \sys_if_shell_restricted: { p , T , F , TF }
+  {
+    \if_int_compare:w \__sys_shell_escape: = 2 \exp_stop_f:
+      \prg_return_true:
+    \else:
+      \prg_return_false:
+    \fi:
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macrocode}
+\tex_chardef:D \c__sys_eighteen = 18 \exp_stop_f:
+% \end{macrocode}
+%
+% \begin{macro}{\sys_shell_now:n}
+%   Execute commands through shell escape immediately.
+%    \begin{macrocode}
+\cs_new_protected:Nn \sys_shell_now:n
+{
+  \sys_if_engine_luatex:TF
+  {
+    \luatex_directlua:D { os.execute(" \luatex_luaescapestring:D { #1 } ") }
+  }
+  {
+    \iow_now:Nn \c__sys_eighteen { #1 }
+  }
+}
+%    \end{macrocode}
+%   \emph{Note:} We could do this fully-expandably in \LuaTeX, but not
+%   in \pdfTeX.  This should be discussed.
+% \end{macro}
+%
+% \begin{macro}{\sys_shell_shipout:n}
+%   Execute commands through shell escape at shipout.
+%    \begin{macrocode}
+\cs_new_protected:Nn \sys_shell_shipout:n
+{
+  \sys_if_engine_luatex:TF
+  {
+    \sys_shell_now:n { #1 }
+  }
+  {
+    \iow_shipout:Nn \c__sys_eighteen { #1 }
+  }
+}
+%    \end{macrocode}
+%   \emph{Note:} There is no method for delay in LuaTeX (perhaps through
+%   a callback).  This should be discussed.
+% \end{macro}
```

Comment:
I don't think you need a callback here just use `\latelua` rather than `\directlua` (cf `\ShellEscape` and `\DelayedShellescape` in tools/shellesc.)